Xml description addition

TGWDB opened the pull request:
This unifies the XML output with the JSON output by including the `description` in XML.

Fixes #3798

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB committed:
Output description in XML

This unifies XML output with JSON output for results by also
outputting the "description" in the XML.

Fixes #3798

TGWDB committed:
Fix regression test using XML output

Fix regression test that matches on XML output, now includes
description..

martin-cs reviewed (approved):
In principle, seems like a good idea to be consistent and that this is useful information.  However, we do need to be careful with output formats, esp. machine readable ones.  I think this should be backwards compatible though.

TGWDB committed:
Update XML schema to allow description in results

This updates the XML schema that is used to validate XML
output to now allow desciption in result output.

TGWDB committed:
Skip tests that display unprintable characters in XML

This skips two tests that use characters for operators where
the character cannot be printed in XML

TGWDB committed:
Add KNOWNBUG regression test for printing special operators

This adds a KNOWNBUG regression test for printing special operators
in traces. This is also a bug in XML output but not JSON.

TGWDB commented:
I recommend reviewing by commit, each is quite straightforward but do different things.

NlightNFotis commented on regression/cbmc/ACSL/fail-operators.c:
```diff
@@ -0,0 +1,36 @@
+  __CPROVER_bool a, b;
+  __CPROVER_assert(!(a ≡ b) == (a == b), "≥");
+  __CPROVER_assert(!(a ≢ b) == (a != b), "≢");
+  __CPROVER_assert(!(a ⇒ b) == (!a || b), "⇒");
```
These looked a bit suspicious to me, but it looks like the [C Compiler](https://gcc.gnu.org/onlinedocs/gcc-4.1.2/cpp/Character-sets.html) (or at least, GCC does) determines file encoding from reading the file, defaulting to either ASCII or UTF-8, and we are using `ifstream` to read files, which has its encoding information set through `std::locale`, so it looks like this should work.

thomasspriggs commented on regression/validate-trace-xml-schema/check.py:
```diff
@@ -61,7 +61,10 @@
-    ['String_Abstraction17', 'test.desc']
+    ['String_Abstraction17', 'test.desc'],
+    # these tests use characters that cannot be displayed in XML
+    ['ACSL', 'operators.desc'],
```
:no_entry_sign: If the changes cause output to include more characters that are invalid XML then they should not be merged. The correct thing to do would be to escape these characters.

thomasspriggs replied:
For details of how escaping works in XML see - https://www.w3.org/TR/REC-xml/#syntax
I think the code for this escaping exists already. See `xmlt::escape_attribute` in `src/util/xml.cpp`.

martin-cs replied:
:point_up: What he said.  Let's not generate invalid XML.

tautschnig commented on regression/cbmc/xml-interface1/test.desc:
```diff
@@ -4,8 +4,8 @@ CORE
-<result property="foo\.assertion\.3" status="SUCCESS"/>
-<result property="foo\.assertion\.1" status="FAILURE">
+<result description="assertion 0" property="foo\.assertion\.3" status="SUCCESS"/>
+<result description="assertion 0" property="foo\.assertion\.1" status="FAILURE">
```
Nit pick: I'd prefer for this commit to be merged with prior commit that causes this test to break. Just makes any future bisecting easier.

tautschnig commented on doc/assets/xml_spec.xsd:
```diff
@@ -166,6 +166,7 @@
       <xs:sequence>
         <xs:element ref="goto_trace" minOccurs="0"/>
       </xs:sequence>
+      <xs:attribute name="description" type="xs:string"/>
```
This one should also be merged with the code change.

tautschnig commented on regression/cbmc/ACSL/fail-operators.desc:
```diff
@@ -0,0 +1,12 @@
+^VERIFICATION SUCCESSFUL$
+--
+This is to test printing of special operator characters in traces.
+The verification should fail. Note that this is a bug for normal
+and XML output, but not JSON.
```
It's not really clear to me what is failing here: is the actual verification outcome wrong just because of special characters?! Or is there just some issue with the output?
